presentation/matrix_view.py

```diff
@@ -159,14 +167,19 @@ def make_sla_levels(self, mesh: MeshResults, metric: MetricType) -> List[SLALeve
             for col in row.columns:
                 warning = thresholds.warning(row.agent_id, col.agent_id)
                 error = thresholds.error(row.agent_id, col.agent_id)
-                value = self.get_metric_value(metric, mesh.connection(row.agent_id, col.agent_id))
-                sla_level = self.get_sla_level(value, warning, error)
+                connection = mesh.connection(row.agent_id, col.agent_id)
+                if connection.is_no_data():
+                    sla_level = SLALevel.NODATA
+                else:
+                    value = self.get_metric_value(metric, connection)
+                    sla_level = self.get_sla_level(value, warning, error)
                 sla_levels_col.append(sla_level)
             sla_levels.append(sla_levels_col)
 
-        for i in range(len(mesh.rows)):  # add Nones at diagonal to avoid colorising it
-            sla_levels[-(i + 1)].insert(i, None)
-
+        # mark matrix diagonal, use alternating SLALevel._MIN and SLALevel._MAX
```

Comment:
This whole value -> level -> color mapping business is getting more and more complicated. I'm starting to think that using simple table (as opposed to the heatmap) is a better solution. It is definitely easier to understand. Visual benefits of the heatmap are marginal and IMHO can be achieved in a simple table too.

Reply:
Also, I do not understand why entries on the diagonal (which obviously have no valid data) should have alternating "sla levels" and how does that simplify construction of the heatmap.

Reply:
We can look for a different component to display the matrix if you think it is worthwhile.
However, heatmap already works and probably won't need much changes in future.
The diagonal is "unused space" in the matrix, so I throw 0.0 and 1.0 there to make sure that min(values in matrix) == 0 and max(values in matrix) ==1, so the color scale matches the matrix content and thus looks correctly.